File: JAVA/Clients/clients/src/main/java/com/futureprograms/clients/controller/ImageController.java
package com.futureprograms.clients.controller;

import jakarta.servlet.http.HttpServletRequest;
import org.springframework.web.servlet.HandlerMapping;
import com.futureprograms.clients.service.ImageService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.bind.annotation.RequestMethod;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * Controlador para servir imágenes subidas por usuarios
 * Maneja la descarga de imágenes de perfil y otros archivos
 */
@RestController
@RequestMapping("/api/images")
@CrossOrigin(
    origins = {"http://localhost", "http://localhost:*", "http://127.0.0.1", "http://localhost:8080", "http://localhost:3000"},
    allowCredentials = "true",
    allowedHeaders = "*",
    methods = {RequestMethod.GET, RequestMethod.OPTIONS},
    maxAge = 3600
)
@Slf4j
@RequiredArgsConstructor
public class ImageController {

    private final ImageService imageService;

    /**
     * Obtiene una imagen por nombre (soporta subcarpetas de usuario como {userId}/profile.jpg)
     * GET /api/images/**
     */
    @GetMapping("/**")
    public ResponseEntity<?> getImage(HttpServletRequest request) {
        // En Spring, si RequestMapping es /api/images y pedimos /api/images/2/profile.jpg
        // PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE suele devolver /2/profile.jpg (la parte variable)
        String path = (String) request.getAttribute(HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE);
        
        // Limpiar la ruta para obtener el nombre relativo real
        String fileName = path;
        // Eliminar cualquier prefijo conocido que pueda venir de la URL o la base de datos
        fileName = fileName.replace("/api/images/", "");
        if (fileName.startsWith("/")) fileName = fileName.substring(1);
        if (fileName.startsWith("images/")) fileName = fileName.substring(7);
        
        // Nueva limpieza: si el nombre viene como "2/profile.jpg" o similar, asegurar que no tenga dobles slashes
        fileName = fileName.replace("//", "/");

        if (fileName.isEmpty() || fileName.equals("health")) {
             return health();
        }

        log.info("🖼️ Solicitud de imagen procesada. Archivo buscado: '{}'", fileName);
        try {
            // Buscar en el directorio de uploads configurado
            // Todas las imágenes (de usuario y por defecto) están centralizadas ahí
            String uploadDirProperty = imageService.getUploadDir();
            Path uploadPath = Paths.get(uploadDirProperty).toAbsolutePath().normalize();
            Path filePath = uploadPath.resolve(fileName).normalize();
            
            log.info("📂 Buscando imagen en: {}", filePath.toAbsolutePath());

            if (Files.exists(filePath) && Files.isReadable(filePath)) {
                log.info("✅ Imagen encontrada");
                Resource resource = new UrlResource(filePath.toUri());
                if (resource.exists() && resource.isReadable()) {
                    return serveResource(resource, filePath);
                }
            }

            log.warn("❌ IMAGEN NO ENCONTRADA: {}", fileName);
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body("{\"error\": \"Imagen no encontrada: " + fileName + "\"}");

        } catch (Exception e) {
            log.error("❌ Error al obtener imagen: {}", fileName, e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Error al obtener la imagen: " + e.getMessage());
        }
    }

    private ResponseEntity<Resource> serveResource(Resource resource, Path path) throws Exception {
        String contentType = Files.probeContentType(path);
        if (contentType == null) {
            contentType = "application/octet-stream";
        }

        log.info("✅ Sirviendo recurso: {} ({})", path.getFileName(), contentType);

        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_TYPE, contentType)
                .header(HttpHeaders.CONTENT_DISPOSITION, "inline; filename=\"" + path.getFileName() + "\"")
                .body(resource);
    }

    /**
     * Health check para servicio de imágenes
     * GET /api/images/health
     */
    @GetMapping("/health")
    public ResponseEntity<?> health() {
        return ResponseEntity.ok()
                .body("{\"status\": \"Image service is running\", \"upload_dir\": \"" + imageService.getUploadDir() + "\"}");
    }
}

File: ASP .NET/Clients/Controllers/ImageController.cs
using Microsoft.AspNetCore.Mvc;
using System.Net;
using Clients.Services;

namespace Clients.Controllers;

/// <summary>
/// Controlador para servir imágenes subidas por usuarios
/// Maneja la descarga de imágenes de perfil y otros archivos
/// 
/// Equivalente al ImageController de Java
/// GET /api/images/{path} - Obtiene imagen por ruta relativa (ej: 1/profile.jpg)
/// </summary>
[ApiController]
[Route("api/images")]
public class ImageController : ControllerBase
{
    private readonly IImageService _imageService;
    private readonly ILogger<ImageController> _logger;
    private readonly IConfiguration _configuration;

    public ImageController(
        IImageService imageService,
        ILogger<ImageController> logger,
        IConfiguration configuration)
    {
        _imageService = imageService;
        _logger = logger;
        _configuration = configuration;
    }

    /// <summary>
    /// Test endpoint para verificar que los archivos estáticos funcionan
    /// GET /api/images/test
    /// </summary>
    [HttpGet("test")]
    public IActionResult TestStaticFiles()
    {
        var uploadDir = _imageService.GetUploadDirectory();
        var user3Dir = Path.Combine(uploadDir, "3");
        var testFile = Path.Combine(user3Dir, "profile.png");

        _logger.LogInformation($"📋 TEST - Upload dir: {uploadDir}");
        _logger.LogInformation($"📋 TEST - User 3 dir: {user3Dir}");
        _logger.LogInformation($"📋 TEST - Test file: {testFile}");
        _logger.LogInformation($"📋 TEST - File exists: {System.IO.File.Exists(testFile)}");

        if (System.IO.File.Exists(testFile))
        {
            var fileInfo = new System.IO.FileInfo(testFile);
            return Ok(new {
                status = "OK",
                file = testFile,
                size = fileInfo.Length,
                exists = true,
                uploadDir = uploadDir,
                accessUrl = "/uploads/3/profile.png"
            });
        }

        return Ok(new {
            status = "File not found",
            file = testFile,
            exists = false,
            uploadDir = uploadDir,
            files = Directory.Exists(user3Dir) ? Directory.GetFiles(user3Dir) : new string[] { }
        });
    }

    /// <summary>
    /// Maneja preflight CORS OPTIONS requests
    /// </summary>
    [HttpOptions("{*path}")]
    public IActionResult OptionsImage(string? path)
    {
        Response.Headers.AccessControlAllowOrigin = "*";
        Response.Headers.AccessControlAllowMethods = "GET, OPTIONS";
        Response.Headers.AccessControlAllowHeaders = "Content-Type";
        Response.Headers.AccessControlMaxAge = "3600";
        return Ok();
    }

    /// <summary>
    /// Obtiene una imagen por ruta relativa
    /// GET /api/images/1/profile.jpg
    /// GET /api/images/default/male.png
    /// GET /api/images/health para verificar el estado del servicio
    /// </summary>
    [HttpGet("{*path}")]
    public async Task<IActionResult> GetImage(string? path)
    {
        try
        {
            // Health check
            if (string.IsNullOrEmpty(path) || path == "health")
            {
                var uploadDirectory = _imageService.GetUploadDirectory();
                _logger.LogInformation($"🏥 Health check - Upload dir: {uploadDirectory}");
                return Ok(new
                {
                    status = "Image service is running",
                    upload_dir = uploadDirectory,
                    timestamp = DateTime.UtcNow
                });
            }

            // List directory
            if (path == "list/3")
            {
                var listDir = _imageService.GetUploadDirectory();
                var userDir = Path.Combine(listDir, "3");
                _logger.LogInformation($"📋 Listando archivos en: {userDir}");
                if (Directory.Exists(userDir))
                {
                    var files = Directory.GetFiles(userDir);
                    return Ok(new { directory = userDir, files = files.Select(f => Path.GetFileName(f)).ToList() });
                }
                return NotFound(new { error = $"Directorio no encontrado: {userDir}" });
            }

            _logger.LogInformation($"🖼️ Solicitud de imagen procesada. Archivo buscado: '{path}'");

            // Limpiar la ruta
            path = path?.Replace("//", "/").TrimStart('/') ?? "";

            if (string.IsNullOrEmpty(path))
            {
                _logger.LogWarning("⚠️ Ruta de imagen vacía");
                return NotFound(new { error = "Ruta de imagen vacía" });
            }

            // Validar que la ruta es segura (no path traversal)
            if (!_imageService.IsValidImagePath(path))
            {
                _logger.LogWarning($"❌ Ruta de imagen inválida: {path}");
                return BadRequest(new { error = "Ruta de imagen inválida" });
            }

            // Obtener el directorio de uploads
            var uploadDir = _imageService.GetUploadDirectory();
            var uploadPath = Path.IsPathRooted(uploadDir)
                ? uploadDir
                : Path.Combine(Directory.GetCurrentDirectory(), uploadDir);

            // Resolver la ruta completa
            var filePath = Path.Combine(uploadPath, path);
            var normalizedPath = Path.GetFullPath(filePath);

            _logger.LogInformation($"📂 Buscando imagen en: {normalizedPath}");
            _logger.LogInformation($"📂 Upload dir: {uploadPath}");
            _logger.LogInformation($"📂 Path original (antes de combine): {path}");

            // Verificar que no es path traversal
            var baseDir = Path.GetFullPath(uploadPath);
            if (!normalizedPath.StartsWith(baseDir + Path.DirectorySeparatorChar) && normalizedPath != baseDir)
            {
                _logger.LogWarning($"❌ Intento de path traversal detectado: {path} | Base: {baseDir} | Normalized: {normalizedPath}");
                return BadRequest(new { error = "Ruta inválida" });
            }

            // Verificar que el archivo existe y es legible
            if (!System.IO.File.Exists(normalizedPath))
            {
                _logger.LogWarning($"❌ Imagen no encontrada: {path} | Full path: {normalizedPath}");
                var parentDir = Path.GetDirectoryName(normalizedPath);
                if (Directory.Exists(parentDir ?? ""))
                {
                    var filesInDir = Directory.GetFiles(parentDir ?? "");
                    _logger.LogWarning($"❌ Archivos en directorio ({parentDir}): {string.Join(", ", filesInDir)}");
                }
                return NotFound(new { error = $"Imagen no encontrada: {path}" });
            }

            _logger.LogInformation($"✅ Imagen encontrada: {path}");

            // Leer el archivo
            var fileBytes = await System.IO.File.ReadAllBytesAsync(normalizedPath);

            // Determinar el content type
            var contentType = GetContentType(normalizedPath);

            _logger.LogInformation($"✅ Sirviendo recurso: {Path.GetFileName(normalizedPath)} ({contentType})");

            // Añadir headers CORS explícitos
            Response.Headers.AccessControlAllowOrigin = "*";
            Response.Headers.AccessControlAllowMethods = "GET, OPTIONS";
            Response.Headers.AccessControlAllowHeaders = "Content-Type";

            // Servir el archivo
            return File(fileBytes, contentType, Path.GetFileName(normalizedPath), enableRangeProcessing: true);
        }
        catch (Exception ex)
        {
            _logger.LogError($"❌ Error al obtener imagen: {path} - {ex.Message}");
            return StatusCode(StatusCodes.Status500InternalServerError,
                new { error = $"Error al obtener la imagen: {ex.Message}" });
        }
    }

    /// <summary>
    /// Determina el content type basado en la extensión del archivo
    /// </summary>
    private string GetContentType(string filePath)
    {
        var ext = Path.GetExtension(filePath).ToLowerInvariant();
        return ext switch
        {
            ".jpg" => "image/jpeg",
            ".jpeg" => "image/jpeg",
            ".png" => "image/png",
            ".gif" => "image/gif",
            ".webp" => "image/webp",
            ".bmp" => "image/bmp",
            _ => "application/octet-stream"
        };
    }
}
